index.d.ts

```diff
@@ -1,288 +1,591 @@
-    disableFilters: boolean
-    setFilter: () => any
-    setAllFilters: () => any
-  }
+export interface UseExpandedRowProps<D extends object> {
+  isExpanded: boolean
+  canExpand: boolean
+  subRows: Row<D>[]
+  toggleExpanded: (isExpanded?: boolean) => void
+  getExpandedToggleProps: (props?: object) => object
+}
+/* #endregion */
 
-  export interface UsePaginationState {
-    pageIndex: number
-    pageSize: number
-    pageCount: number
-    rowCount: number
-  }
+/* #region useFilters */
+export function useFilters<D extends object = {}>(hooks: Hooks<D>): void
+export namespace useFilters {
+  const pluginName = 'useFilters'
+}
 
-  export interface UsePaginationValues<D = {}> {
-    page: Page<D>
-    pageIndex: number // yes, this is on instance and state
-    pageSize: number // yes, this is on instance and state
-    canPreviousPage: boolean
-    canNextPage: boolean
-    nextPage: () => any
-    previousPage: () => any
-    setPageSize: (size: number) => any
-    pageOptions: any[]
-    manualPagination: boolean
-    paginateExpandedRows: boolean
-    disablePageResetOnDataChange: boolean
-    pageCount: number
-    gotoPage: (page: number) => any
-  }
+export type UseFiltersOptions<D extends object> = Partial<{
+  manualFilters: boolean
+  disableFilters: boolean
+  filterTypes: Filters<D>
+}>
 
-  export interface UseFiltersState<D> {
-    filters?: Filters<D>
-  }
+export interface UseFiltersState<D extends object> {
+  filters: Filters<D>
+}
 
-  export interface UseFiltersValues<D> {
-    setFilter: (columnID: keyof D, value: string) => void
-    setAllFilters: (values: Filters<D>) => void
-    disableFilters: boolean
-  }
+export type UseFiltersColumnOptions<D extends object> = Partial<{
+  disableFilters: boolean
+  Filter: Renderer<FilterProps<D>>
+  filter: FilterType<D> | DefaultFilterTypes | keyof Filters<D> // TODO: Check that this doesn't lose the DefaultFilterTypes intellisense
+}>
+
+export interface UseFiltersInstanceProps<D extends object> {
+  rows: Row<D>[]
+  preFilteredRows: Row<D>[]
+  setFilter: (
+    columnId: IdType<D>,
+    updater: ((filterValue: FilterValue) => FilterValue) | FilterValue // NOTE: `| FilterValue` makes this evaluate to `any`
+  ) => void
+  setAllFilters: (
+    updater: Filters<D> | ((filters: Filters<D>) => Filters<D>)
+  ) => void
+}
 
-  export interface UseGroupByValues {
-    groupByFn: any
-    manualGroupBy: boolean
-    disableGrouping: boolean
-    aggregations: any
-  }
+export interface UseFiltersColumnProps<D extends object> {
+  canFilter: boolean
+  setFilter: (
+    updater: ((filterValue: FilterValue) => FilterValue) | FilterValue
+  ) => void
+  filterValue: FilterValue
+  preFilteredRows: Row<D>[]
+}
 
-  export interface UseGroupByState {
-    groupBy: string[]
-    isAggregated?: boolean
-  }
+export type FilterProps<D extends object> = HeaderProps<D>
+export type FilterValue = any
+export type Filters<D extends object> = Record<IdType<D>, FilterValue>
```

Comment:
should this be something like

```tsx
export type Filters<D extends object> = Record<IdType<D>, FilterValue> | {}
```

since it's valid to pass in an empty object to unset all filters, in fact I'm not sure how else to unset them all.

Reply:
On a `Record` you can pass an empty object. I don't see a benefit in adding the union.
https://www.typescriptlang.org/play/#code/C4TwDgpgBAYglgG2BATgZygXigJQgYwHsUATAHjWBTgDsBzAGigEMaQA+AKG6JspYBcsRMnRYoAbwC+QA

Reply:
you can, but without a cast you get:

```tsx
TS2345: Argument of type '{}' is not assignable to parameter of type 'Record<IdType<T>, any> | ((filters: Record<IdType<T>, any>) => Record<IdType<T>, any>)'.
  Type '{}' is not assignable to type '(filters: Record<IdType<T>, any>) => Record<IdType<T>, any>'.
    Type '{}' provides no match for the signature '(filters: Record<IdType<T>, any>): Record<IdType<T>, any>'.

100       setAllFilters({})
```

You can fix it with:

```tsx
      setAllFilters({} as Record<IdType<T>,any>)
```

But I thought that simply allowing the empty object was cleaner and didn't require the user to use casts.

Reply:
I'm not able to reproduce this. I dropped `setAllFilters({})` into my kitchen sink example and it works just fine. Can you provide a reproducible case?

Reply:
huh, same problem here.  Error in one case but not in the other.  I'm guessing that this will come down to some difference in the tsconfig, my production test has several things switched off :(  and I've seen that cause issues in the past.

Since this works correctly with the default strict settings, I think that this is good as is.